esda/correlogram.py

```diff
@@ -0,0 +1,153 @@
+    for attribute in all_attrs:
+        attrs.append(getattr(autocorr, str(attribute)))
+    return pd.Series(attrs, index=all_attrs, name=dist)
+
+
+def correlogram(
+    gdf: gpd.GeoDataFrame,
+    variable: str,
+    distances: list,
+    distance_type: str = "band",
+    statistic: str = "I",
+    weights_kwargs: dict = None,
+    stat_kwargs: dict = None,
+    n_jobs: int = -1,
+    backend: str = "loky",
+):
+    """Generate a spatial correlogram
+
+    A spatial correlogram is a set of spatial autocorrelation statistics calculated for
+    a set of increasing distances. It is a useful exploratory tool for examining
+    how the relationship between spatial units changes over different notions of scale.
+
+    Parameters
+    ----------
+    gdf : gpd.GeoDataFrame
+        geodataframe holding spatial and attribute data
+    variable: str
+        column on the geodataframe used to compute autocorrelation statistic
+    distances : list
+        list of distances to compute the autocorrelation statistic
+    distance_type : str, optional
+        which concept of distance to increment. Options are {`band`, `knn`}.
+        by default 'band' (for `libpysal.weights.DistanceBand` weights)
+    statistic : str, by default 'I'
+        which spatial autocorrelation statistic to compute. Options in {`I`, `G`, `C`}
+    weights_kwargs : dict
+        additional keyword arguments passed to the libpysal.weights.W class
+    stat_kwargs : dict
+        additional keyword arguments passed to the `esda` autocorrelation statistic class.
+        For example for faster results with no statistical inference, set the number
+        of permutations to zero with {permutations: 0}
+    n_jobs : int
+        number of jobs to pass to joblib. If -1 (default), all cores will be used
+    backend : str
+        backend parameter passed to joblib
+
+    Returns
+    -------
+    outputs : pandas.DataFrame
+        table of autocorrelation statistics at increasing distance bandwidths
+    """
+    if stat_kwargs is None:
+        stat_kwargs = dict()
+    if weights_kwargs is None:
+        weights_kwargs = dict()
+    if statistic == "I":
+        STATISTIC = Moran
+    elif statistic == "G":
+        STATISTIC = G
+    elif statistic == "C":
+        STATISTIC = Geary
+    else:
+        with NotImplementedError as e:
+            raise e("Only I, G, and C statistics are currently implemented")
+
+    if distance_type == "band":
+        W = DistanceBand
+    elif distance_type == "knn":
+        if max(distances) > gdf.shape[0] - 1:
+            with ValueError as e:
+                raise e("max number of neighbors must be less than or equal to n-1")
+        W = KNN
+    else:
+        with NotImplementedError as e:
+            raise e("distance_type must be either `band` or `knn` ")
+
+    #  should be able to build the tree once and reuse it?
+    #  but in practice, im not seeing any real difference from starting a new W from scratch each time
```

Comment:
Not even for large data? I suppose that repeated creation of the tree could make a dent on this.

Reply:
I was "testing" on a dataset with ~65k points, but just not terribly thoroughly

what i meant was, I'd also done `KNN.from_dataframe(df, k=k)` in each loop, instead of `W(tree)`, where the former needs to build the tree each time in addition to querying it. Then i figured it might be faster to re-use the same tree repeatedly instead

what i was hoping for was something like @ljwolf said, where i'd do the expensive call once for the largest distance, then repeatedly use the same structure for repeated calls (sorta like how pandana works with `preprocess`). If we just loop over k sorted in descending order, does that help?

Reply:
(a) I think it's fine to keep this current implementation for now. 
(b) yeah basically: a more performant implementation would probably use `max(distances)` to compute `.sparse_distance_matrix()` (or built a sparse matrix from the output of `query()` if KNN) one time, and then progressively mask that based on each smaller distance value... basically, we want to query once, rather than repeat the query to yield results we can "precompute".

Reply:
gotcha. [Same way](https://github.com/oturns/geosnap/blob/main/geosnap/analyze/network.py) i did it for the isochrones with pandana. Get the adjacency once and trim it down for each distance. So here it would be something like 

```python

maxtree = KDTree(max(distances))
mintree = KDTree(min(distances))

distmat = maxtree.sparse_distance_matrix(mintree)
for dist in distances
	w = Graph.build_knn(distmat, k=k)
```
or something?

Reply:
i guess what im asking is, what's the most efficient way to do this with the new Graph? 

If we have a Graph based on the max distance, then it's already got all the information we need for every lesser distance, and all we need to do is filter on the adjlist where wij<distance. Presumably thats what simething like `Graph.build_knn(distmat, k=k)` would do when passed an adjlist or a sparse matrix because all it needs to do is subset the existing data, not rebuild the distance relationships

for this problem its easy enough to just build the adjlist once and literally filter it down each distance and instantiate a new Graph/W from that adjlist, but is that the best way?

Reply:
> for this problem its easy enough to just build the adjlist once and literally filter it down each distance and instantiate a new Graph/W from that adjlist, but is that the best way?

I suppose this would need to be tested. My sense is that a filter of adjacency based on a distance will be significantly slower than a tree query but I might be wrong.

Reply:
so, i'm playing with this, and it would be straightforward to filter the adjlist, which is fast, since its just `adj[adj['weight']<=distance]`. 

...But once we subset the W, now we have to subset the original dataframe to align it, which is something we said we wouldnt do. Is there a good way to do this?

Reply:
To keep the structure, you can just assign weight=0 to everything above the set distance threshold.